### PR TITLE
feat(linter/eslint-plugin-promise): implement valid-params

### PR DESCRIPTION
**Pull request by jelly**
Rule detail:
[link](https://github.com/eslint-community/eslint-plugin-promise/blob/main/docs/rules/valid-params.md)

---

Split this up from https://github.com/oxc-project/oxc/pull/4318 to make it easier to review and as some of the other rules require some more complicated work.

### CODE REVIEW

File: crates/oxc_linter/src/rules/promise/valid_params.rs [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+    // hello.then(), hello.catch(), hello.finally()
+    if matches!(prop_name, "then" | "catch" | "finally") {
+        return true;
+    }
+
+    member_expr.object().is_specific_id("Promise")
+        && matches!(
+            prop_name,
+            "resolve" | "reject" | "all" | "allSettled" | "race" | "any" | "withResolvers"
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/promise/valid_params.rs:50]: Should we have a shared utility for promise names that can be used by other rules? We may also want to use `phf_set`.

File: crates/oxc_linter/src/rules/promise/valid_params.rs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let AstKind::CallExpression(call_expr) = node.kind() else {
+            return;
+        };
+
+        if !is_promise(call_expr) {
+            return;
+        }
+
+        let Some(member_expr) = call_expr.callee.get_member_expr() else {
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/promise/valid_params.rs:64]: These two checks are redundant with logic in is_promise. Maybe have that function return the references you need?

Comment (jelly) [reply crates/oxc_linter/src/rules/promise/valid_params.rs:64]: Seems logical, I checked the other PR and at least `no-return-in-finallly` could re-use that, other rules can't but they can instead call `is_promse().is_some()`

File: crates/oxc_linter/src/rules/promise/valid_params.rs (hunk 1 above)
Comment (jelly) [reply crates/oxc_linter/src/rules/promise/valid_params.rs:50]: I think I do that in another PR.

https://github.com/oxc-project/oxc/pull/4318/files#diff-0afbabe6273a5175aa7ba4cf2f9282b1b6318e738ab5d6b65b12ae831a469414R5

So  I will sneak that into this one :)

File: crates/oxc_linter/src/rules/promise/valid_params.rs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        let Some(prop_name) = is_promise(call_expr) else {
+            return;
+        };
+
+        let args_len = call_expr.arguments.len();
+
+        match prop_name.as_str() {
+            "resolve" | "reject" => {
+                if args_len > 1 {
+                    ctx.diagnostic(valid_params_diagnostic(call_expr.span, &format!("Promise.{prop_name}() requires 0 or 1 arguments, but received {args_len}")));
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/promise/valid_params.rs:48]: Move message formatting logic into the diagnostics function

Comment (DonIsaac) [reply crates/oxc_linter/src/rules/promise/valid_params.rs:48]: Also, this code creates a String, passes a &str to the diagnostics function, which then creates another String. This results in a duplicate string allocation. The above suggestion fixes this.

File: crates/oxc_linter/src/rules/promise/valid_params.rs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+                }
+            }
+            "then" => {
+                if !(1..=2).contains(&args_len) {
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/promise/valid_params.rs:52]: While cleaner, using a range here creates an iterator, which produces unnecessary branches. Consider just using a greater than and less than or equals check.

File: crates/oxc_linter/src/snapshots/valid_params.snap [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+  ⚠ eslint-plugin-promise(valid-params): Promise.allSettled() requires 1 argument, but received 4
+   ╭─[valid_params.tsx:1:1]
+ 1 │ Promise.allSettled({}, function() {}, 1, 2)
+   · ───────────────────────────────────────────
+   ╰────
+
+  ⚠ eslint-plugin-promise(valid-params): Promise.any() requires 1 argument, but received 3
+   ╭─[valid_params.tsx:1:1]
+ 1 │ Promise.any(1, 2, 3)
```
Comment (DonIsaac) [inline crates/oxc_linter/src/snapshots/valid_params.snap:66]: Could we add a suggestion for this case? Using `ctx.diagnostic_with_suggestion`?

File: crates/oxc_linter/src/rules/promise/valid_params.rs (hunk 4 above)
Comment (jelly) [reply crates/oxc_linter/src/rules/promise/valid_params.rs:52]: Sure, I have to figure out how to silence clippy then.

File: crates/oxc_linter/src/snapshots/valid_params.snap (hunk 5 above)
Comment (jelly) [reply crates/oxc_linter/src/snapshots/valid_params.snap:66]: Hmm well the arguments can be any type so I haven't figured out how to get the span location yet of the first argument. Or use the source text to determine the location of the comma

File: crates/oxc_linter/src/rules/promise/valid_params.rs [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+                        args_len,
+                    ));
+                }
+            }
+            _ => {}
+        }
+    }
+}
+
+// fn generate_fix<'a>(
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/promise/valid_params.rs:112]: looks like it needs deleting

**Comment by DonIsaac** [comment]
One last thing, then we're good to go 🎊

File: crates/oxc_linter/src/rules/promise/valid_params.rs (hunk 6 above)
Comment (jelly) [reply crates/oxc_linter/src/rules/promise/valid_params.rs]: Sorry, dropped it.

**Review by DonIsaac — APPROVED** [review]

**Pull request merged**